### ISSUE TITLE
High CPU use on add_waker

### ISSUE THREAD

**Issue by sjmiller609** [geofmureithi/apalis#477]
I've noticed a high polling rate with PostgreSQL

```
                                                       query                                                        | calls |  total_exec_time  |    avg_exec_time    | avg_rows |     hit_percent
--------------------------------------------------------------------------------------------------------------------+-------+-------------------+---------------------+----------+----------------------
 Select * from apalis.get_jobs($1, $2, $3)                                                                          |  3946 | 673.3816409999986 | 0.17064917410035443 |        0 |  99.7274692228174044
 UPDATE apalis.jobs                                                                                                +|  3942 | 211.1570709999998 | 0.05356597437848803 |        0 | 100.0000000000000000
                             SET status = $4, done_at = $5, lock_by = $6, lock_at = $7, last_error = $8            +|       |                   |                     |          |
                             WHERE id IN                                                                           +|       |                   |                     |          |
                                 (SELECT jobs.id FROM apalis.jobs INNER JOIN apalis.workers ON lock_by = workers.id+|       |                   |                     |          |
                                     WHERE status = $9                                                             +|       |                   |                     |          |
                                     AND workers.last_seen < ($3::timestamp)                                       +|       |                   |                     |          |
                                     AND workers.worker_type = $1                                                  +|       |                   |                     |          |
                                     ORDER BY lock_at ASC                                                          +|       |                   |                     |          |
                                     LIMIT $2)                                                                      |       |                   |                     |          |
(2 rows)
```

[This repository](https://github.com/sjmiller609/reproduce-issue/tree/main) can help reproduce the issue.

**Comment by geofmureithi** [comment]
We had this issue https://github.com/geofmureithi/apalis/pull/472 and this should be resolved by the new release v0.6.1. This was caused by the way CallAll is designed in tower.
Could you try the same profiling with the latest version?

**Comment by sjmiller609** [comment]
Thank you for getting back so quickly. I've updated to 0.6.1, and thank you for the suggestion. However, I think I'm still having a high polling rate. When I re-run the test, I'm getting 384 database requests over a 10 second period of sampling.

```
                                                       query                                                        | calls |  total_exec_time  |    avg_exec_time    | avg_rows |     hit_percent
--------------------------------------------------------------------------------------------------------------------+-------+-------------------+---------------------+----------+----------------------
 Select * from apalis.get_jobs($1, $2, $3)                                                                          |  2618 | 380.5825580000012 | 0.14537148892284232 |        0 |  99.8307692307692308
 UPDATE apalis.jobs                                                                                                +|  2616 | 105.9006530000001 | 0.04048190099388383 |        0 | 100.0000000000000000
                             SET status = $4, done_at = $5, lock_by = $6, lock_at = $7, last_error = $8            +|       |                   |                     |          |
                             WHERE id IN                                                                           +|       |                   |                     |          |
                                 (SELECT jobs.id FROM apalis.jobs INNER JOIN apalis.workers ON lock_by = workers.id+|       |                   |                     |          |
                                     WHERE status = $9                                                             +|       |                   |                     |          |
                                     AND workers.last_seen < ($3::timestamp)                                       +|       |                   |                     |          |
                                     AND workers.worker_type = $1                                                  +|       |                   |                     |          |
                                     ORDER BY lock_at ASC                                                          +|       |                   |                     |          |
                                     LIMIT $2)                                                                      |       |                   |                     |          |
(2 rows)

.... 10 seconds later
                                                       query                                                        | calls |  total_exec_time   |    avg_exec_time    | avg_rows |     hit_percent
--------------------------------------------------------------------------------------------------------------------+-------+--------------------+---------------------+----------+----------------------
 Select * from apalis.get_jobs($1, $2, $3)                                                                          |  2810 |   405.886464000001 | 0.14444358149466227 |        0 |  99.8402091807088902
 UPDATE apalis.jobs                                                                                                +|  2808 | 113.03406800000008 | 0.04025429772079775 |        0 | 100.0000000000000000
                             SET status = $4, done_at = $5, lock_by = $6, lock_at = $7, last_error = $8            +|       |                    |                     |          |
                             WHERE id IN                                                                           +|       |                    |                     |          |
                                 (SELECT jobs.id FROM apalis.jobs INNER JOIN apalis.workers ON lock_by = workers.id+|       |                    |                     |          |
                                     WHERE status = $9                                                             +|       |                    |                     |          |
                                     AND workers.last_seen < ($3::timestamp)                                       +|       |                    |                     |          |
                                     AND workers.worker_type = $1                                                  +|       |                    |                     |          |
                                     ORDER BY lock_at ASC                                                          +|       |                    |                     |          |
                                     LIMIT $2)                                                                      |       |                    |                     |          |
(2 rows)

```

**Background information:**

In my linked reproduction repo, the CPU is not very hot because 384 queries over 10 seconds isn't that fast when there are no other queries in the database. In my server, the workload that has other things going on in it (writes to other relations of the database for example), I find CPU growing over time. I have added a temporary workaround to 1) isolate apalis in its own container and 2) restart the container every hour.

(version 0.6.0)

![image](https://github.com/user-attachments/assets/6dafe453-98b4-46d7-a892-80b066ac2211)


CPU profiling indicated this is the hot code section. (version 0.6.0)

https://github.com/geofmureithi/apalis/blob/094938d95f7ad90caab9edfc9fd04368fbcbe193/packages/apalis-core/src/worker/mod.rs#L493-L499

![image](https://github.com/user-attachments/assets/06398a3a-89c1-4b99-beb1-cbdc9a943c82)

The longer it was left running without a restart in a nearly-idle state (2 jobs per minute), the greater percentage the flame graph will show over time in that code section. This indicates this code section is where the growing CPU time is being spent, seen as the sawtooth graph (first picture in this post).

Then, I worked on trying to isolate the problem in [this](https://github.com/sjmiller609/reproduce-issue) repository. I could not reproduce the hot flame graph on add_waker. During the investigation, I noticed this is possibly connected to a high polling rate, which I discovered by turning on trace-level logs in apalis core, then measuring with pg_stat_statements.

Yesterday, before filing this issue, I tried running from **main** of apalis, but I still had the hot / growing CPU in the server and high polling rate to the database.

**Version 0.6.1 (today)**

First, I have updated the reproduction repo, and it is seeing a high polling rate (see top of this post).

Next, I have updated my server to 0.6.1 (was running on main branch as of yesterday overnight with the problem still happening). I started the server with 0.6.1 at 1:34pm ET. Near to starting the server, we can see add_waker and deeper in the stack at 3.51% of CPU time.

![image](https://github.com/user-attachments/assets/48c5c79b-5f58-43e9-bcca-25e3ee7707d0)

After waiting 2 minutes, it's grown to 16.74% of the time.

![image](https://github.com/user-attachments/assets/42c51032-609f-403d-ac7d-8fc1a27f8cd7)

After a few more minutes, still growing.

![image](https://github.com/user-attachments/assets/0096698a-a98d-40f8-ba7b-965ffabb7a27)

The jobs container is increasing in CPU utilization over time. There is only 1 job running per minute, and it's completing successfully, showing as "Done" in the DB, and the logs look fine.

I think my problem is still happening on 0.6.1.

CPU utilization over time, 0.6.1

![image](https://github.com/user-attachments/assets/516c46fe-4b28-4dc8-8b43-259266d9ea4e)


Thank you.

**Comment by geofmureithi** [comment]
I need to dive deeper into this, but one more check you can do is tinker around with the configs:
```
impl Default for Config {
    fn default() -> Self {
        Self {
            keep_alive: Duration::from_secs(30),
            buffer_size: 10,
            poll_interval: Duration::from_millis(100),
            reenqueue_orphaned_after: Duration::from_secs(300), // 5 minutes
            namespace: String::from("apalis::sql"),
        }
    }
}
```
I think the 100 milliseconds poll interval is too low.

Meanwhile, let me look through the add_waker issue, seems we have a bug somewhere.

**Comment by sjmiller609** [comment]
I've had the server running with the polling interval configured to 1 second (up from 100ms) for a couple hours.

The CPU appears not growing so far, you can see where the sawtooth function stops, that's when I updated the configuration.

![image](https://github.com/user-attachments/assets/6102420a-3fcb-4cd5-9588-305f929b661f)

However, I still see a large percent of the time on the add_waker function. This is surprising because I thought the slower poll fixed the issue, based on the CPU seeming flat.

![image](https://github.com/user-attachments/assets/381e108a-f0a3-4423-81b3-51a57220371f)

The high percentage on add_waker and growing over time CPU utilization only applies to my server, I haven't got this to show up in the reproduction repository.

Because I still see the add_waker in the profile still, I've tried zooming in on the recent CPU:

![image](https://github.com/user-attachments/assets/b2666d54-7f40-49af-9fbb-067e972dbe50)

The Y axis is percent CPU utilization, 0 to 100. This is the container running only apalis jobs, nearly idle with 1 job running per minute. We can see the CPU utilization of the jobs is actually still increasing, but just very slowly. So the polling change made a big difference on how fast the CPU grows, but did not completely fix the issue. It is good enough for my use case though.

In summary, this polling adjustment has made a big difference, with CPU utilization growing much more slowly, but it still seems like something is leaking to cause CPU to grow on the add_waker function.

**Comment by geofmureithi** [comment]
@sjmiller609 I have this [PR](https://github.com/geofmureithi/apalis/pull/481) that attempts to resolve the issue. 
The add waker function now just adds the latest wakable waker. Let me know how this goes for you.

**Comment by sjmiller609** [comment]
Awesome, I think that worked! Thank you. https://github.com/geofmureithi/apalis/pull/481#issuecomment-2539753252
